File: Large_File_Processing/java/StreamAggregate.java
// Constant-memory aggregation over a huge FundsXML file via StAX (native Java,
// no JAXB, no DOM). Pull parser — never holds more than the current element.
//
//   javac -d /tmp/lf Large_File_Processing/java/StreamAggregate.java
//   java  -cp /tmp/lf StreamAggregate <fundsxml.xml>
//
// FundsXML 4.x has no XML namespace — match the bare "Position" local name.
// Security: external entities / DTDs disabled (XXE-safe).

import java.io.FileInputStream;
import javax.xml.stream.XMLInputFactory;
import javax.xml.stream.XMLStreamConstants;
import javax.xml.stream.XMLStreamReader;

public class StreamAggregate {
    public static void main(String[] args) throws Exception {
        if (args.length != 1) {
            System.err.println("usage: StreamAggregate <fundsxml.xml>");
            System.exit(2);
        }

        XMLInputFactory xif = XMLInputFactory.newInstance();
        xif.setProperty(XMLInputFactory.IS_SUPPORTING_EXTERNAL_ENTITIES, false);
        xif.setProperty(XMLInputFactory.SUPPORT_DTD, false);

        long n = 0;
        double sumValue = 0, sumPct = 0;
        boolean inPosition = false, inTotalValue = false;
        String currentLeaf = null;

        try (FileInputStream in = new FileInputStream(args[0])) {
            XMLStreamReader r = xif.createXMLStreamReader(in);
            while (r.hasNext()) {
                switch (r.next()) {
                    case XMLStreamConstants.START_ELEMENT: {
                        String name = r.getLocalName();
                        if ("Position".equals(name)) { inPosition = true; n++; }
                        else if ("TotalValue".equals(name)) inTotalValue = true;
                        currentLeaf = name;
                        break;
                    }
                    case XMLStreamConstants.CHARACTERS: {
                        if (!inPosition || r.isWhiteSpace()) break;
                        String t = r.getText().trim();
                        if (t.isEmpty()) break;
                        if (inTotalValue && "Amount".equals(currentLeaf))
                            sumValue += Double.parseDouble(t);
                        else if ("TotalPercentage".equals(currentLeaf))
                            sumPct += Double.parseDouble(t);
                        break;
                    }
                    case XMLStreamConstants.END_ELEMENT: {
                        String name = r.getLocalName();
                        if ("Position".equals(name)) inPosition = false;
                        else if ("TotalValue".equals(name)) inTotalValue = false;
                        currentLeaf = null;
                        break;
                    }
                    default:
                }
            }
            r.close();
        }

        Runtime rt = Runtime.getRuntime();
        long usedMiB = (rt.totalMemory() - rt.freeMemory()) / (1024 * 1024);
        System.out.printf("positions      : %d%n", n);
        System.out.printf("sum value (EUR): %.2f%n", sumValue);
        System.out.printf("sum percentage : %.4f%n", sumPct);
        System.out.printf("heap in use    : %d MiB%n", usedMiB);
    }
}

File: Large_File_Processing/python/stream_aggregate.py
#!/usr/bin/env python3
"""Constant-memory aggregation over a huge FundsXML file via lxml iterparse.

Usage:  stream_aggregate.py <fundsxml.xml>

Streams Position elements one at a time, summing value (fund ccy) and
percentage and counting positions, then frees each element AND its already-
processed previous siblings so memory stays flat regardless of file size
(the classic lxml fast_iter pattern). Prints the totals and peak RSS.

FundsXML 4.x has no XML namespace — match the bare 'Position' tag.

Dependencies: lxml (`pip install lxml`); Python stdlib otherwise.
Security: iterparse runs with resolve_entities=False, no_network=True and
huge_tree=False — XXE / entity-expansion safe even on untrusted feeds.
"""
import resource
import sys

from lxml import etree


def main() -> int:
    if len(sys.argv) != 2:
        print("usage: stream_aggregate.py <fundsxml.xml>", file=sys.stderr)
        return 2
    path = sys.argv[1]

    n = 0
    sum_value = 0.0
    sum_pct = 0.0

    # Pull events for Position end-tags only; no DTD/entity expansion.
    context = etree.iterparse(path, events=("end",), tag="Position",
                              resolve_entities=False, no_network=True,
                              huge_tree=False)
    for _, pos in context:
        amt = pos.find("./TotalValue/Amount")
        if amt is not None and amt.text:
            sum_value += float(amt.text)
        p = pos.findtext("TotalPercentage")
        if p:
            sum_pct += float(p)
        n += 1
        # Free this element and earlier siblings already parsed.
        pos.clear()
        while pos.getprevious() is not None:
            del pos.getparent()[0]
    del context

    peak_kb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    print(f"positions      : {n}")
    print(f"sum value (EUR): {sum_value:.2f}")
    print(f"sum percentage : {sum_pct:.4f}")
    print(f"peak RSS       : {peak_kb // 1024} MiB")
    return 0


if __name__ == "__main__":
    sys.exit(main())
